### PR TITLE
Add placeholders for console support that explain how to do it

### PR DESCRIPTION
**Pull request by Calinou**
The Xbox and PlayStation icon background colors were taken from their respective official websites.

Supersedes #40194. This pull request uses the official console logos and adds an Xbox placeholder export preset.

## Preview

(I can't display the actual notice due to single-window popups being broken :slightly_frowning_face:)

image

<details>
<summary>Old version</summary>

image
</details>

### CODE REVIEW

File: editor/editor_export.h [hunk 1]
```diff
@@ -483,4 +485,46 @@ class EditorExportTextSceneToBinaryPlugin : public EditorExportPlugin {
...
+	}
+	virtual String get_os_name() const override {
+		return console_os_name;
+	}
+	virtual Ref<Texture2D> get_logo() const override {
+		return logo;
+	}
+
+	bool can_export(const Ref<EditorExportPreset> &p_preset, String &r_error, bool &r_missing_templates, bool &r_missing_console) const override {
+		r_error = vformat(TTR("Exporting to %s requires getting licensed with %s, then obtaining export templates from a third-party company (or porting it yourself)."), console_name, console_licensing);
```
Comment (reduz) [inline editor/editor_export.h:509]: I was thinking about:

As %s does not authorize public distribution of their SDKs, you need to get licensed with them first. Afterwards, export templates are commercially available from third party companies. More information here...

Comment (Calinou) [reply editor/editor_export.h:509]: I can't check how the message looks in practice, so I need to check whether it'll wrap automatically or not. If not, I'll need to use `String::word_wrap`.

**Comment by breadnone** [comment]
This would give user fake hopes and sorry to say this but it looks like a bogus marketing scheme. The idea is to attract users and not to discourage them of exporting the game themselves.

Reasons:
1. Requires to be licensed <- means cost
2. Requires template from 3rd party <- means another "DOUBLE the costs"
**3.** At this point, you're already penniless and broke, probably living on the streets somewhere! 

Question remains!
Why the templates should be provided from a 3rd party!?

**Comment by Calinou** [comment]
> Why the templates should be provided from a 3rd party!?

We can't offer 1st-party templates as the Software Freedom Conservancy (the nonprofit managing Godot's funds) doesn't allow us to distribute proprietary software.

**Comment by groud** [comment]
I think the idea is good. However I agree a little bit with @breadnone as it kind of give the user false hopes.
To solve this, I would gray out the entries in the export list, so that it's obvious those are not available "out of the box" and should be enabled via thirdparties.

**Comment by Calinou** [comment]
@groud If we disable the entries, we can't click them anymore. This would make it impossible to see the help message.

**Comment by groud** [comment]
> @groud If we disable the entries, we can't click them anymore. This would make it impossible to see the error message.

Ah that's not the best solution then. :/
Maybe the icons should be grayed out then? Like, converted to black & white?

**Comment by Calinou** [comment]
> Maybe the icons should be grayed out then? Like, converted to black & white?

Sounds good to me.

**Comment by breadnone** [comment]
> It's why there's a disclaimer about porting to consoles being expensive. I wouldn't consider porting your first game to a console unless it's already sold well elsewhere 🙂
> 
Sure, BUT this only applicable for more mature audience but not the younger ones. Young mindset would easily targetting "console" as their go to platform, I bet my ear on this. 

There are two possibility that could come from this pr:
1. It would stop them from the get go, obviously, because of no console support.
2. Makes them despise the engine for not straightforwardly telling it upfront.

I'd like to see Godot getting bigger, but not like this

**Comment by bruvzg** [comment]
> Maybe the icons should be grayed out then? Like, converted to black & white?

Gray it out, cross it out and add info about licensing right to the menu items.

Screenshot 2020-09-27 at 15 02 45

> I need to check whether it'll wrap automatically or not before changing the message.

It's not, window width will expand to fit full message instead.

Unrelated note: Export icons ignore `EDSCALE` and are tiny on the hiDPI display.

**Comment by groud** [comment]
> cross it out

Usually, such cross mean more that something is forbidden than unavailable. So I believe it is not necessary here.
But otherwise, mentioning "Third-party license required" is a good idea.

**Comment by Calinou** [comment]
@groud Done:

image

File: editor/editor_export.h (hunk 1 above)
Comment (bruvzg) [reply editor/editor_export.h:508]: It does not word wrap automatically but respect `\n` in the text, current line is a bit too long as well.

Screenshot 2020-09-27 at 19 18 16

```suggestion
		r_error = vformat(TTR("As %s does not authorize public distribution of their SDKs, you need to get licensed with them first.\nAfterwards, export templates are commercially available from third party companies."), console_licensing);
```

Screenshot 2020-09-27 at 19 22 19

**Comment by wkubiak** [comment]
@Calinou are you sure that it's legal to use these logos as part of a software product that's not tied/licensed to the trademark and platform holders?

I have serious doubts, so I'd advise to check up on the legal documents of each platform (or at least ask support via email) before making these changes public.

**Comment by Listwon** [comment]
Neither Unity nor Unreal Engine shows Nintendo Switch on their platform exports lists. Actually Unreal doesn't show any consoles there unless you are licensed and have access to corresponding exporters. So why would you add this to the core of open-source engine when you are not licensed?

**Comment by groud** [comment]
> So why would you add this to the core of open-source engine when you are not licensed?

Because people keep thinking that Godot is not capable to export to consoles, while it already is (thanks to thirdparties).
This is a question that comes up so often that it is better for us if we can tell users directly in the editor that they can eventually export to console.

**Comment by Listwon** [comment]
@groud The question was about the legal part, not the reason behind this. And it's just cluttering interface with unusable stuff.

**Comment by starry-abyss** [comment]
An alternative could be a button "Where are the other platforms?" linking to Godot website's platforms hub, which would say where to get unofficial support for consoles, Raspberry Pi, etc.

**Comment by SIsilicon** [comment]
I have a suggestion! How about whenever the console buttons are pressed, this shows up.
image

**Comment by Calinou** [comment]
> I have a suggestion! How about whenever the console buttons are pressed, this shows up.

The message would have to be reworded since there is technically no build of the editor that can export to consoles. Also, the "Ok" button should open the URL using `OS.shell_open()` since text in Godot dialogs usually can't be copied.

**Comment by SIsilicon** [comment]
> The message would have to be reworded since there is technically no build of the editor that can export to consoles.

Technically there _are_ builds of the editor that can export to consoles. They're just not _official_ builds, and are only owned by those with access to the console SDKs.

**Comment by SIsilicon** [comment]
> As far as I know, what third party companies provide you are console export templates, not custom editor builds.

Really? How are the templates used if the editor doesn't even know how to use them?

**Comment by SIsilicon** [comment]
> Also, the OK button should be renamed to "Open in Browser" and should open the URL using `OS.shell_open()`. This is because text in Godot dialogs usually can't be copied. (I'd change it to a ConfirmationDialog so you can cancel too.)

Like this?
image

**Comment by Calinou** [comment]
> Like this?

Yes :slightly_smiling_face:

**Comment by reduz** [comment]
Okay, what do we do with this?

**Comment by groud** [comment]
> Okay, what do we do with this?

Personally, the fact this solution might give users false hopes made me think we should maybe go for a simpler solution.

Maybe, instead of having entries for each console, maybe we could simply put a link at the bottom of the list (not inside it) to the documentation page?  Probably something like "Exporting to consoles or other proprietary hardware...". That way we would still have a link there but we do not make users think export templates are easily available and provided as simple add-ons.

**Comment by wkubiak** [comment]
I'm with @groud on this one. Let's put one common "Consoles" item in the menu and that will open the popup. Much less clutter and no false hopes.

**Comment by mhilbrunner** [comment]
I'm closing this due to:

1. There not being consensus of whether this should be done, or whether its misleading.
2. This use of the logos likely not being legal:
    1. They may allow the logo being used to refer to compatability (but we're really showing _in_compatability, which usually isn't allowed explicitely; whether this falls under the former is likely a legal matter)
    2. They _don't_ allow the logo being altered (e.g. turned to grayscale).
    3. In the case of the Playstation and Switch logos, both Nintendo and Sony have multiple additional license/brand requirements which we would be breaching with this but I can't share those publically due to NDA, and I can't find similar documents online to what I have available locally.

This also inconsistently refers to only the Nintendo Switch as the _Nintendo_ Switch, while the others are called "PlayStation" (not "Sony PlayStation") and "Xbox" (not "Microsoft Xbox"). The former is also a specific console, the latter are console families.

Refer to https://www.microsoft.com/en-us/legal/intellectualproperty/trademarks

**Comment by Zireael07** [comment]
@mhilbrunner: Nintendo Switch vs Playstation and Xbox is probably due to the fact that "Switch" is a common word, not just this one product's name.

And re: turning the logo grayscale, are they really THIS anal?

**Pull request closed without merging**